netty/src/main/java/io/grpc/netty/NettyChannelBuilder.java

```diff
@@ -259,8 +264,19 @@ private static String getAuthorityFromAddress(SocketAddress address) {
    * your application won't start.
    */
   @CanIgnoreReturnValue
+  @SuppressWarnings("unchecked")
   public NettyChannelBuilder channelType(Class<? extends Channel> channelType) {
     checkNotNull(channelType, "channelType");
+    try {
+      Method method = channelType.getMethod("remoteAddress");
+      checkNotNull(method, "method for remoteAddress");
+      Class<?> returnType = method.getReturnType();
+      checkState(SocketAddress.class.isAssignableFrom(returnType),
+          "Unexpected return type:" + returnType.getSimpleName());
+      this.transportSocketType = (Class<? extends SocketAddress>) returnType;
```

Comment:
This only mutates one of the code paths. If the user calls `channelFactory()` directly, then the address types get in an inconsistent state. For the factory, it seems we need users to pass in the supported types for us to do any checking.

I feel like we can't fix up all cases now (or maybe ever), so there needs to be an opt-out of the address compat check. That could be "return null from getSupportedSocketAddressTypes()". We'd use it in channelFactory() and if vanilla SocketAddress is detected here. And we'd add a new overload of channelType()/channelFactory() where you pass in the supported types to skip the reflection. We'd then use that overload in UdsNettyChannelBuilder.

Reply:
Addressed.